File: 02/gpt-5.1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strings"
)

func main() {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	input := strings.TrimSpace(string(data))

	part1 := sumInvalidIDs(input)
	fmt.Printf("Day 02 Part 1: %d\n", part1)
	part2 := sumInvalidIDsPart2(input)
	fmt.Printf("Day 02 Part 2: %d\n", part2)
}

func sumInvalidIDs(input string) int64 {
	sum := int64(0)
	ranges := strings.Split(input, ",")
	for _, r := range ranges {
		r = strings.TrimSpace(r)
		if r == "" {
			continue
		}
		parts := strings.Split(r, "-")
		if len(parts) != 2 {
			continue
		}
		start, err1 := parseInt64(strings.TrimSpace(parts[0]))
		end, err2 := parseInt64(strings.TrimSpace(parts[1]))
		if err1 != nil || err2 != nil {
			continue
		}
		if end < start {
			start, end = end, start
		}

		minLen := digitsCount(start)
		maxLen := digitsCount(end)
		for l := minLen; l <= maxLen; l++ {
			if l%2 != 0 {
				continue
			}
			half := l / 2
			pow10 := pow10Int64(half)
			first := pow10Int64(l - 1)
			last := pow10Int64(l) - 1

			low := maxInt64(start, first)
			high := minInt64(end, last)
			if low > high {
				continue
			}

			den := float64(pow10 + 1)
			aStart := int64(math.Ceil(float64(low) / den))
			aEnd := int64(math.Floor(float64(high) / den))
			if aStart > aEnd {
				continue
			}

			count := aEnd - aStart + 1
			termsSum := (aStart + aEnd) * count / 2
			contrib := termsSum * (pow10 + 1)
			sum += contrib
		}
	}
	return sum
}

func sumInvalidIDsPart2(input string) int64 {
	sum := int64(0)
	ranges := strings.Split(input, ",")
	for _, r := range ranges {
		r = strings.TrimSpace(r)
		if r == "" {
			continue
		}
		parts := strings.Split(r, "-")
		if len(parts) != 2 {
			continue
		}
		start, err1 := parseInt64(strings.TrimSpace(parts[0]))
		end, err2 := parseInt64(strings.TrimSpace(parts[1]))
		if err1 != nil || err2 != nil {
			continue
		}
		if end < start {
			start, end = end, start
		}

		minLen := digitsCount(start)
		maxLen := digitsCount(end)
		for l := minLen; l <= maxLen; l++ {
			for k := 2; k*k <= l; k++ {
				if l%k != 0 {
					continue
				}
				baseLen := l / k
				pow10Base := pow10Int64(baseLen)
				first := pow10Int64(l - 1)
				last := pow10Int64(l) - 1

				low := maxInt64(start, first)
				high := minInt64(end, last)
				if low > high {
					continue
				}

				// geometric series factor: 1 + 10^{baseLen} + ...
				geom := int64(0)
				factor := int64(1)
				for i := 0; i < k; i++ {
					geom += factor
					factor *= pow10Base
				}

				aStart := int64(math.Ceil(float64(low) / float64(geom)))
				aEnd := int64(math.Floor(float64(high) / float64(geom)))
				if aStart > aEnd {
					continue
				}

				count := aEnd - aStart + 1
				termsSum := (aStart + aEnd) * count / 2
				contrib := termsSum * geom
				sum += contrib
			}
		}
	}
	return sum
}

func parseInt64(s string) (int64, error) {
	reader := strings.NewReader(s)
	scanner := bufio.NewScanner(reader)
	scanner.Split(bufio.ScanWords)
	if scanner.Scan() {
		var v int64
		_, err := fmt.Sscan(scanner.Text(), &v)
		return v, err
	}
	return 0, fmt.Errorf("invalid int64")
}

func digitsCount(n int64) int {
	if n < 0 {
		n = -n
	}
	count := 1
	for n >= 10 {
		n /= 10
		count++
	}
	return count
}

func pow10Int64(exp int) int64 {
	res := int64(1)
	for i := 0; i < exp; i++ {
		res *= 10
	}
	return res
}

func maxInt64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func minInt64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

File: 02/gpt-5.1/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt")
        .expect("Failed to read input")
        .trim()
        .to_string();

    let part1 = sum_invalid_ids(&input);
    println!("Day 02 Part 1: {}", part1);

    let part2 = sum_invalid_ids_part2(&input);
    println!("Day 02 Part 2: {}", part2);
}

fn sum_invalid_ids(input: &str) -> i64 {
    let mut sum: i64 = 0;
    for range in input.split(',') {
        let r = range.trim();
        if r.is_empty() {
            continue;
        }
        let parts: Vec<&str> = r.split('-').map(|s| s.trim()).collect();
        if parts.len() != 2 {
            continue;
        }
        let mut start: i64 = match parts[0].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let mut end: i64 = match parts[1].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        if end < start {
            std::mem::swap(&mut start, &mut end);
        }

        let min_len = digits_count(start);
        let max_len = digits_count(end);
        for len in min_len..=max_len {
            if len % 2 != 0 {
                continue;
            }

            let half = len / 2;
            let pow10 = pow10_i64(half);
            let first = pow10_i64(len - 1);
            let last = pow10_i64(len) - 1;

            let low = std::cmp::max(start, first);
            let high = std::cmp::min(end, last);
            if low > high {
                continue;
            }

            let den = (pow10 + 1) as f64;
            let a_start = (low as f64 / den).ceil() as i64;
            let a_end = (high as f64 / den).floor() as i64;
            if a_start > a_end {
                continue;
            }

            let count = a_end - a_start + 1;
            let terms_sum = (a_start + a_end) * count / 2;
            let contrib = terms_sum * (pow10 + 1);
            sum += contrib;
        }
    }

    sum
}

fn sum_invalid_ids_part2(input: &str) -> i64 {
    let mut sum: i64 = 0;
    for range in input.split(',') {
        let r = range.trim();
        if r.is_empty() {
            continue;
        }
        let parts: Vec<&str> = r.split('-').map(|s| s.trim()).collect();
        if parts.len() != 2 {
            continue;
        }
        let mut start: i64 = match parts[0].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let mut end: i64 = match parts[1].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        if end < start {
            std::mem::swap(&mut start, &mut end);
        }

        let min_len = digits_count(start);
        let max_len = digits_count(end);
        for len in min_len..=max_len {
            for k in 2.. {
                if k * k > len { break; }
                if len % k != 0 { continue; }

                let base_len = len / k;
                let pow10_base = pow10_i64(base_len);
                let first = pow10_i64(len - 1);
                let last = pow10_i64(len) - 1;

                let low = std::cmp::max(start, first);
                let high = std::cmp::min(end, last);
                if low > high { continue; }

                let mut geom: i64 = 0;
                let mut factor: i64 = 1;
                for _i in 0..k {
                    geom += factor;
                    factor *= pow10_base;
                }

                let a_start = ((low as f64) / (geom as f64)).ceil() as i64;
                let a_end = ((high as f64) / (geom as f64)).floor() as i64;
                if a_start > a_end { continue; }

                let count = a_end - a_start + 1;
                let terms_sum = (a_start + a_end) * count / 2;
                let contrib = terms_sum * geom;
                sum += contrib;
            }
        }
    }

    sum
}

fn digits_count(mut n: i64) -> i32 {
    if n < 0 {
        n = -n;
    }
    let mut count = 1;
    while n >= 10 {
        n /= 10;
        count += 1;
    }
    count
}

fn pow10_i64(exp: i32) -> i64 {
    let mut res: i64 = 1;
    for _ in 0..exp {
        res *= 10;
    }
    res
}
